components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/internal/action/PreUpdateProfileActionExecutor.java

```diff
@@ -0,0 +1,304 @@
+            }
+        }
+    }
+
+    private void populateUpdatingClaimsInUserActionRequestDTO(Map<String, String> userClaims,
+                                                              UserActionRequestDTO.Builder userActionRequestDTOBuilder)
+            throws UserStoreException {
+
+        String multiAttributeSeparator = FrameworkUtils.getMultiAttributeSeparator();
+        for (Map.Entry<String, String> entry : userClaims.entrySet()) {
+            String claimKey = entry.getKey();
+            String claimValue = entry.getValue();
+            if (isMultiValuedClaim(claimKey)) {
+                userActionRequestDTOBuilder.addClaim(claimKey,
+                        StringUtils.isBlank(claimValue) ? new String[]{} :
+                                claimValue.split(Pattern.quote(multiAttributeSeparator)));
+            } else {
+                userActionRequestDTOBuilder.addClaim(claimKey, claimValue);
+            }
+        }
+    }
+
+    /**
+     * Populate the multi-valued claims to be modified by adding the values to be added and
+     * removing the values to be removed from the existing claim values of a particular multi-valued claim of a user.
+     *
+     * @param existingClaimsOfUser      Existing claims of the user.
+     * @param multiValuedClaimsToAdd    Multi-valued claims to be added.
+     * @param multiValuedClaimsToDelete Multi-valued claims to be deleted.
+     * @return Multi-valued claims to be modified.
+     */
+    private Map<String, String> getSimpleMultiValuedClaimsToModify(
+            Map<String, String> existingClaimsOfUser,
+            Map<String, List<String>> multiValuedClaimsToAdd,
+            Map<String, List<String>> multiValuedClaimsToDelete) {
+
+        String multiAttributeSeparator = FrameworkUtils.getMultiAttributeSeparator();
+        Map<String, String> multiValuedClaimsToModify = new HashMap<>();
+
+        Map<String, List<String>> existingClaimsAsList = existingClaimsOfUser.entrySet().stream()
+                .collect(Collectors.toMap(
+                        Map.Entry::getKey,
+                        entry -> new ArrayList<>(
+                                Arrays.asList(entry.getValue().split(Pattern.quote(multiAttributeSeparator))))));
+
+        Function<String, List<String>> getMutableClaimList = claimURI -> {
+            if (multiValuedClaimsToModify.containsKey(claimURI)) {
+                return new ArrayList<>(Arrays.asList(
+                        multiValuedClaimsToModify.get(claimURI).split(Pattern.quote(multiAttributeSeparator))));
+            }
+            return new ArrayList<>(existingClaimsAsList.getOrDefault(claimURI, new ArrayList<>()));
+        };
+
+        // Add the values to be added to the existing values of each multi-valued claim and
+        // construct the new claim value joining the values
+        if (multiValuedClaimsToAdd != null) {
+            multiValuedClaimsToAdd.forEach((claimURI, valuesToAdd) -> {
```

Comment:
Shouldn't we check, that multiValuedClaimsToAdd and multiValuedClaimsToDelete are actually multi-value claims. Is it already validating somewhere else?

Reply:
Also what will happen, if extension service send the multi-valued claim in userClaimsExcludingMultiValuedClaimsToBeModified and vice versa.

Reply:
SCIM layer already resolves exact values to be added and removed from the value set of a multi attribute claim and this executor executes on top of that

Reply:
To answer the second question, as of now we have not dealt with letting the extension to modify claim sets. This needs to be sorted in next phase